test/test_packet.py

```diff
@@ -53,77 +67,81 @@ def loopback_test(self, dw):
-                        dut.last_errors += 1
-            yield
+            datas = [prng.randrange(2**8) for _ in range(prng.randrange(dw - 1) + 1)]
+            packets.append(StreamPacket(datas, header))
 
         class DUT(Module):
             def __init__(self):
-                packetizer   = Packetizer(packet_description(dw), raw_description(dw), packet_header)
-                depacketizer = Depacketizer(raw_description(dw), packet_description(dw), packet_header)
-                self.submodules += packetizer, depacketizer
-                self.comb += packetizer.source.connect(depacketizer.sink)
-                self.sink, self.source = packetizer.sink, depacketizer.source
+                self.submodules.packetizer = Packetizer(
+                    packet_description(dw),
+                    raw_description(dw),
+                    packet_header,
+                )
+                self.submodules.depacketizer = Depacketizer(
+                    raw_description(dw),
+                    packet_description(dw),
+                    packet_header,
+                )
+                self.comb += self.packetizer.source.connect(self.depacketizer.sink)
+                self.sink, self.source = self.packetizer.sink, self.depacketizer.source
 
         dut = DUT()
-        run_simulation(dut, [generator(dut), checker(dut)])
-        self.assertEqual(dut.header_errors, 0)
-        self.assertEqual(dut.data_errors,   0)
-        self.assertEqual(dut.last_errors,   0)
+        recvd_packets = []
+        run_simulation(
+            dut,
+            [
+                stream_inserter(
+                    dut.sink,
+                    src=packets,
+                    seed=seed,
+                    debug_print=debug_print,
+                    valid_rand=50,
+                ),
+                stream_collector(
+                    dut.source,
+                    dest=recvd_packets,
+                    expect_npackets=npackets,
+                    seed=seed,
+                    debug_print=debug_print,
+                    ready_rand=50,
+                ),
+            ],
+            vcd_name="{}.vcd".format(with_last_be),
+        )
+
+        # When we don't have a last_be signal, the Packetizer will
+        # simply throw away the partial bus word. The Depacketizer
+        # will then fill up these values with garbage again. Thus we
+        # also have to remove the proper amount of bytes from the sent
+        # packets so the comparison will work.
+        if not with_last_be and dw != 8:
+            for (packet, recvd_packet) in zip(packets, recvd_packets):
+                invalid_recvd_bytes = packet_header_length % (dw // 8)
+                recvd_packet.data = recvd_packet.data[:-invalid_recvd_bytes]
+                packet.data = packet.data[:len(recvd_packet.data)]
+
+        self.assertTrue(compare_packets(packets, recvd_packets))
```

Comment:
One thing to note is that the Packetizer, without any `last_be` signal, will throw away the last partial bus word. For instance, when sending `0xDEAD BEEF BADC 0DED` over a 32-bit data path with a 3-byte header of `0xABC`, the following packet will be sent: `0xABCD EADB EEFB ADC0`. This can be quite unintuitive. The Depacketizer will then restore the `0xABC` header, producing the data `0xDEAD BEEF BADC 0XXX` where X are unrelated signal states.

Whether this should change (so that all data plus some trailing garbage would be encoded) is a different question, but the above logic handles these cases accordingly.

Reply:
Adding randomly generated header fields to the test suite revealed that this statement is not true unconditionally. If the `header_leftover` (number of bytes of the header which are interleaved with data in a single bus word) plus the number of bytes in the last aligned data word is below or equal to the number of bytes in a bus word, all data (plus some trailing garbage) will be transmitted. Example:

Header: `0xABCD E`
Data: `0xBADD EADD AD`
Packetized: `0xABCD EBAD DEAD DADx` (where x is are unrelated bytes)
Depacketized: Header `0xABCD E`, Data `0xBADD EADD ADxx` (where x are unrelated bytes)

Hence I've adjusted the test logic.

Reply:
Just an idea,. Instead of trying to guess/fake the last_be signal, could this generate a warning/error if the last_be signal isn't present? That way the user can determine the behavior they want?

Reply:
I think that, yes, it might be reasonable to refuse operation without a `last_be` signal with dw > 8. However, that would be a breaking change and I think it'd be preferable to have this in its own PR. I also don't know whether any subsystems rely on the current behavior -- which this PR doesn't change, except for 1-2 bug fixes.

Also, I really don't have much more time for this, and given it does now work correctly as far as I can tell I don't see a reason per sé to deliberate disable otherwise functional code paths. :smile: Generating a warning does sound very reasonable though, @enjoy-digital what do you think?